File: go/tests/retail_board_test.go
// retail_board_test.go
//
// Verifies the CircleAI.Retail port (retail_board.go): product add/get, stock
// set/get, RecordSale (append + stock decrement, unknown-SKU error), RevenueToday
// (same-day sum of UnitPrice*Quantity), and TopSellersSince (units-descending,
// topK cap, first-seen tie order).

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestRetail_ProductsStockAndSale(t *testing.T) {
	b := circleai.NewInMemoryRetailBoard()
	b.AddProduct(circleai.Product{Sku: "A", Name: "Widget", Price: circleai.DecimalFromInt(10), Currency: "ZAR"})
	b.SetStock(circleai.StockLevel{Sku: "A", Quantity: 100})

	if p, ok := b.GetProduct("A"); !ok || p.Name != "Widget" {
		t.Fatalf("get product = %+v ok=%v", p, ok)
	}
	if b.Stock("A") != 100 {
		t.Fatalf("stock = %d, want 100", b.Stock("A"))
	}
	if b.Stock("Z") != 0 {
		t.Fatalf("unknown stock must be 0")
	}

	now := time.Date(2026, 7, 10, 12, 0, 0, 0, time.UTC)
	if err := b.RecordSale(circleai.Sale{SaleId: "s1", Sku: "A", Quantity: 3, UnitPrice: circleai.DecimalFromInt(10), AtUtc: now}); err != nil {
		t.Fatalf("record sale: %v", err)
	}
	if b.Stock("A") != 97 {
		t.Fatalf("stock after sale = %d, want 97", b.Stock("A"))
	}
	// Unknown SKU errors.
	if err := b.RecordSale(circleai.Sale{SaleId: "s2", Sku: "GHOST", Quantity: 1, UnitPrice: circleai.DecimalFromInt(1), AtUtc: now}); err == nil {
		t.Fatalf("unknown SKU sale must error")
	}
}

func TestRetail_RevenueToday(t *testing.T) {
	b := circleai.NewInMemoryRetailBoard()
	b.AddProduct(circleai.Product{Sku: "A", Name: "A", Price: circleai.DecimalFromInt(10), Currency: "ZAR"})
	now := time.Date(2026, 7, 10, 12, 0, 0, 0, time.UTC)
	yesterday := now.AddDate(0, 0, -1)
	_ = b.RecordSale(circleai.Sale{SaleId: "s1", Sku: "A", Quantity: 2, UnitPrice: circleai.DecimalFromInt(10), AtUtc: now})
	_ = b.RecordSale(circleai.Sale{SaleId: "s2", Sku: "A", Quantity: 5, UnitPrice: circleai.DecimalFromInt(10), AtUtc: now.Add(2 * time.Hour)})
	_ = b.RecordSale(circleai.Sale{SaleId: "s3", Sku: "A", Quantity: 9, UnitPrice: circleai.DecimalFromInt(10), AtUtc: yesterday})

	// Today: (2 + 5) * 10 = 70. Yesterday's 90 excluded.
	if rev := b.RevenueToday(now); !rev.Equal(circleai.DecimalFromInt(70)) {
		t.Fatalf("revenue today = %s, want 70", rev)
	}
}

func TestRetail_TopSellers(t *testing.T) {
	b := circleai.NewInMemoryRetailBoard()
	for _, sku := range []string{"A", "B", "C"} {
		b.AddProduct(circleai.Product{Sku: sku, Name: sku, Price: circleai.DecimalFromInt(1), Currency: "ZAR"})
	}
	base := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	since := base.Add(-time.Hour)
	_ = b.RecordSale(circleai.Sale{SaleId: "s1", Sku: "A", Quantity: 5, UnitPrice: circleai.DecimalFromInt(1), AtUtc: base})
	_ = b.RecordSale(circleai.Sale{SaleId: "s2", Sku: "B", Quantity: 8, UnitPrice: circleai.DecimalFromInt(1), AtUtc: base})
	_ = b.RecordSale(circleai.Sale{SaleId: "s3", Sku: "A", Quantity: 4, UnitPrice: circleai.DecimalFromInt(1), AtUtc: base})                   // A -> 9 total
	_ = b.RecordSale(circleai.Sale{SaleId: "s4", Sku: "C", Quantity: 1, UnitPrice: circleai.DecimalFromInt(1), AtUtc: base.AddDate(0, 0, -5)}) // before `since`, excluded

	top, err := b.TopSellersSince(since, 5)
	if err != nil {
		t.Fatalf("top sellers: %v", err)
	}
	// A=9, B=8 (C excluded by time). Descending: A then B.
	if len(top) != 2 || top[0].Sku != "A" || top[0].Sold != 9 || top[1].Sku != "B" || top[1].Sold != 8 {
		t.Fatalf("top sellers wrong: %+v", top)
	}
	// topK cap.
	if one, _ := b.TopSellersSince(since, 1); len(one) != 1 || one[0].Sku != "A" {
		t.Fatalf("topK cap wrong: %+v", one)
	}
	if _, err := b.TopSellersSince(since, 0); err == nil {
		t.Fatalf("topK<=0 must error")
	}
}

File: typescript/tests/retail_board.test.ts
// retail_board.test.ts
// Verifies the CircleAI.Retail port: products + stock, sale recording with
// stock decrement + unknown-SKU guard, same-day revenue, top-sellers ranking.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { InMemoryRetailBoard, RetailDomainContext, product, stockLevel, sale } from "../src/retail/index";

const D = (s: string) => new Date(s);

describe("InMemoryRetailBoard", () => {
  it("adds products, sets and reads stock (missing = 0)", () => {
    const b = new InMemoryRetailBoard();
    b.addProduct(product("A", "Apple", 5, "ZAR", "Fruit"));
    b.setStock(stockLevel("A", 20));
    assert.equal(b.getProduct("A")?.name, "Apple");
    assert.equal(b.stock("A"), 20);
    assert.equal(b.stock("missing"), 0);
  });

  it("records sales, decrements stock, and rejects unknown SKUs", () => {
    const b = new InMemoryRetailBoard();
    b.addProduct(product("A", "Apple", 5, "ZAR", "Fruit"));
    b.setStock(stockLevel("A", 20));
    b.recordSale(sale("s1", "A", 3, 5, D("2026-05-01T10:00:00Z")));
    assert.equal(b.stock("A"), 17);
    assert.throws(() => b.recordSale(sale("s2", "ZZZ", 1, 1, D("2026-05-01T10:00:00Z"))), /Unknown SKU ZZZ/);
  });

  it("sums revenue for the given UTC day only", () => {
    const b = new InMemoryRetailBoard();
    b.addProduct(product("A", "Apple", 5, "ZAR", "Fruit"));
    b.recordSale(sale("s1", "A", 2, 5, D("2026-05-01T08:00:00Z"))); // today
    b.recordSale(sale("s2", "A", 1, 5, D("2026-05-01T23:00:00Z"))); // today
    b.recordSale(sale("s3", "A", 4, 5, D("2026-04-30T23:00:00Z"))); // other day
    assert.equal(b.revenueToday(D("2026-05-01T12:00:00Z")), 15); // (2+1)*5
  });

  it("ranks top sellers since a cutoff, take topK", () => {
    const b = new InMemoryRetailBoard();
    b.addProduct(product("A", "Apple", 5, "ZAR", null));
    b.addProduct(product("B", "Banana", 3, "ZAR", null));
    b.recordSale(sale("s1", "A", 2, 5, D("2026-05-01T10:00:00Z")));
    b.recordSale(sale("s2", "B", 9, 3, D("2026-05-02T10:00:00Z")));
    b.recordSale(sale("s3", "A", 4, 5, D("2026-05-03T10:00:00Z")));
    b.recordSale(sale("s4", "B", 1, 3, D("2026-04-01T10:00:00Z"))); // before cutoff
    const top = b.topSellersSince(D("2026-05-01T00:00:00Z"), 5);
    assert.deepEqual(top, [
      { sku: "B", sold: 9 },
      { sku: "A", sold: 6 },
    ]);
    assert.equal(b.topSellersSince(D("2026-05-01T00:00:00Z"), 1).length, 1);
    assert.throws(() => b.topSellersSince(D("2026-05-01T00:00:00Z"), 0));
  });

  it("rejects null arguments", () => {
    const b = new InMemoryRetailBoard();
    assert.throws(() => b.addProduct(null as never));
    assert.throws(() => b.setStock(null as never));
    assert.throws(() => b.recordSale(null as never));
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(RetailDomainContext.systemPromptSnippet.includes("[DOMAIN: Retail]"));
    assert.deepEqual(RetailDomainContext.complianceFlags, ["Consumer_Protection_Act", "POPIA", "Labour_Relations_Act"]);
    assert.deepEqual(RetailDomainContext.suggestedTools, ["pos_system", "inventory", "analytics", "promotions_engine"]);
  });
});
